# **Level 3**
## **Task 1: Predictive Modeling**

In [107]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
data=pd.read_csv(r"cleaned_dataset.csv")
data.head()

Restaurant ID         Restaurant Name  Country Code              City  \
0        6317637        Le Petit Souffle           162       Makati City   
1        6304287        Izakaya Kikufuji           162       Makati City   
2        6300002  Heat - Edsa Shangri-La           162  Mandaluyong City   
3        6318506                    Ooma           162  Mandaluyong City   
4        6314302             Sambo Kojin           162  Mandaluyong City   

                                             Address  \
0  Third Floor, Century City Mall, Kalayaan Avenu...   
1  Little Tokyo, 2277 Chino Roces Avenue, Legaspi...   
2  Edsa Shangri-La, 1 Garden Way, Ortigas, Mandal...   
3  Third Floor, Mega Fashion Hall, SM Megamall, O...   
4  Third Floor, Mega Atrium, SM Megamall, Ortigas...   

                                     Locality  \
0   Century City Mall, Poblacion, Makati City   
1  Little Tokyo, Legaspi Village, Makati City   
2  Edsa Shangri-La, Ortigas, Mandaluyong City   
3      SM Megamall, Ortigas, Mandaluyong City   
4      SM Megamall, Ortigas, Mandaluyong City   

                                    Locality Verbose   Longitude   Latitude  \
0  Century City Mall, Poblacion, Makati City, Mak...  121.027535  14.565443   
1  Little Tokyo, Legaspi Village, Makati City, Ma...  121.014101  14.553708   
2  Edsa Shangri-La, Ortigas, Mandaluyong City, Ma...  121.056831  14.581404   
3  SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.056475  14.585318   
4  SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.057508  14.584450   

                           Cuisines  ...          Currency Has Table booking  \
0        French, Japanese, Desserts  ...  Botswana Pula(P)               Yes   
1                          Japanese  ...  Botswana Pula(P)               Yes   
2  Seafood, Asian, Filipino, Indian  ...  Botswana Pula(P)               Yes   
3                   Japanese, Sushi  ...  Botswana Pula(P)                No   
4                  Japanese, Korean  ...  Botswana Pula(P)               Yes   

  Has Online delivery Is delivering now Switch to order menu Price range  \
0                  No                No                   No           3   
1                  No                No                   No           3   
2                  No                No                   No           4   
3                  No                No                   No           4   
4                  No                No                   No           4   

   Aggregate rating  Rating color Rating text Votes  
0               4.8    Dark Green   Excellent   314  
1               4.5    Dark Green   Excellent   591  
2               4.4         Green   Very Good   270  
3               4.9    Dark Green   Excellent   365  
4               4.8    Dark Green   Excellent   229  

[5 rows x 21 columns]

In [108]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9551 entries, 0 to 9550
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Restaurant ID         9551 non-null   int64  
 1   Restaurant Name       9551 non-null   object 
 2   Country Code          9551 non-null   int64  
 3   City                  9551 non-null   object 
 4   Address               9551 non-null   object 
 5   Locality              9551 non-null   object 
 6   Locality Verbose      9551 non-null   object 
 7   Longitude             9551 non-null   float64
 8   Latitude              9551 non-null   float64
 9   Cuisines              9542 non-null   object 
 10  Average Cost for two  9551 non-null   int64  
 11  Currency              9551 non-null   object 
 12  Has Table booking     9551 non-null   object 
 13  Has Online delivery   9551 non-null   object 
 14  Is delivering now     9551 non-null   object 
 15  Switch to order menu 

In [109]:
data.isnull().sum()

Restaurant ID           0
Restaurant Name         0
Country Code            0
City                    0
Address                 0
Locality                0
Locality Verbose        0
Longitude               0
Latitude                0
Cuisines                9
Average Cost for two    0
Currency                0
Has Table booking       0
Has Online delivery     0
Is delivering now       0
Switch to order menu    0
Price range             0
Aggregate rating        0
Rating color            0
Rating text             0
Votes                   0
dtype: int64

In [110]:
import warnings
warnings.filterwarnings('ignore')
for i in data:
    if data[i].dtypes=='object':
        mode=data[i].mode()[0]
        data[i].fillna(mode,inplace=True)
    else:
        mean=round((data[i].mean()),3)
        data[i].fillna(mean,inplace=True)
data.isnull().sum()

Restaurant ID           0
Restaurant Name         0
Country Code            0
City                    0
Address                 0
Locality                0
Locality Verbose        0
Longitude               0
Latitude                0
Cuisines                0
Average Cost for two    0
Currency                0
Has Table booking       0
Has Online delivery     0
Is delivering now       0
Switch to order menu    0
Price range             0
Aggregate rating        0
Rating color            0
Rating text             0
Votes                   0
dtype: int64

### Removing Unique,ID,unwanted  columns:

In [111]:
for i in data:
    print(f'{i}-{len(data[i].unique())}')

Restaurant ID-9551
Restaurant Name-7446
Country Code-15
City-141
Address-8918
Locality-1208
Locality Verbose-1265
Longitude-8120
Latitude-8677
Cuisines-1825
Average Cost for two-140
Currency-12
Has Table booking-2
Has Online delivery-2
Is delivering now-2
Switch to order menu-1
Price range-4
Aggregate rating-33
Rating color-6
Rating text-6
Votes-1012


- **Restaurant ID** can be remove because it is an ID column
- **Resturant Name** can be remove because it is mostly a unique column
- **Address** column can be removed because of too large address ,we can use city instead 
- **Longitude,Latitude**  column can be removed because of  large numerical values,  we can use city instead
- **Locality,Locality Verbose** columns can be removed ,we can use city instead
- **Switch to order menu**` can be removed because it has only one unique value Throughout the dataset

In [112]:
data=data.drop(['Restaurant ID','Restaurant Name','Address','Longitude','Latitude','Switch to order menu','Locality','Locality Verbose'],axis=1)
data.head()

Country Code              City                          Cuisines  \
0           162       Makati City        French, Japanese, Desserts   
1           162       Makati City                          Japanese   
2           162  Mandaluyong City  Seafood, Asian, Filipino, Indian   
3           162  Mandaluyong City                   Japanese, Sushi   
4           162  Mandaluyong City                  Japanese, Korean   

   Average Cost for two          Currency Has Table booking  \
0                  1100  Botswana Pula(P)               Yes   
1                  1200  Botswana Pula(P)               Yes   
2                  4000  Botswana Pula(P)               Yes   
3                  1500  Botswana Pula(P)                No   
4                  1500  Botswana Pula(P)               Yes   

  Has Online delivery Is delivering now  Price range  Aggregate rating  \
0                  No                No            3               4.8   
1                  No                No            3               4.5   
2                  No                No            4               4.4   
3                  No                No            4               4.9   
4                  No                No            4               4.8   

  Rating color Rating text  Votes  
0   Dark Green   Excellent    314  
1   Dark Green   Excellent    591  
2        Green   Very Good    270  
3   Dark Green   Excellent    365  
4   Dark Green   Excellent    229

### **3.1.1 Build a regression model to predict the aggregate rating of a restaurant based on available features.**

In [113]:
data['Country Code']=data['Country Code'].astype('object')
data['Price range']=data['Price range'].astype('object')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9551 entries, 0 to 9550
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Country Code          9551 non-null   object 
 1   City                  9551 non-null   object 
 2   Cuisines              9551 non-null   object 
 3   Average Cost for two  9551 non-null   int64  
 4   Currency              9551 non-null   object 
 5   Has Table booking     9551 non-null   object 
 6   Has Online delivery   9551 non-null   object 
 7   Is delivering now     9551 non-null   object 
 8   Price range           9551 non-null   object 
 9   Aggregate rating      9551 non-null   float64
 10  Rating color          9551 non-null   object 
 11  Rating text           9551 non-null   object 
 12  Votes                 9551 non-null   int64  
dtypes: float64(1), int64(2), object(10)
memory usage: 970.2+ KB


In [114]:
cat=data.select_dtypes(include='object').columns
num=data.select_dtypes(exclude='object').columns
cat,num

(Index(['Country Code', 'City', 'Cuisines', 'Currency', 'Has Table booking',
        'Has Online delivery', 'Is delivering now', 'Price range',
        'Rating color', 'Rating text'],
       dtype='object'),
 Index(['Average Cost for two', 'Aggregate rating', 'Votes'], dtype='object'))

### Encoding Categorical columns
- Applying Normal encoding for all categorical columns except 'Cuisines'
- Applying **MultiLabelBinarizer** for **'Cuisines'** column because 'Cuisines' columns consist of multiple v 

In [115]:
from sklearn.preprocessing import MultiLabelBinarizer
enc=list(cat)

# Step 1: Multi-Label Encoding for "Cuisines"
data['Cuisines'] = data['Cuisines'].str.split(', ')  # Convert string to list
mlb = MultiLabelBinarizer()
cuisine_encoded = pd.DataFrame(mlb.fit_transform(data['Cuisines']), columns=mlb.classes_)
data=data.drop('Cuisines', axis=1).join(cuisine_encoded)  # Merge encoded cuisines
enc.remove( 'Cuisines')

# Step 2: Encoding Other Categorical Columns
for i in enc:
    d={}
    k=0
    unique=data[i].unique()
    if 'Yes' in unique or 'No' in unique:
        d['Yes']=1
        d['No']=0
    elif i=='Price range':
        continue
    else:
        for j in unique:
            d[j]=k
            k+=1
    data[i]=data[i].map(d)
    print(d)


{162: 0, 30: 1, 216: 2, 14: 3, 37: 4, 184: 5, 214: 6, 1: 7, 94: 8, 148: 9, 215: 10, 166: 11, 189: 12, 191: 13, 208: 14}
{'Makati City': 0, 'Mandaluyong City': 1, 'Pasay City': 2, 'Pasig City': 3, 'Quezon City': 4, 'San Juan City': 5, 'Santa Rosa': 6, 'Tagaytay City': 7, 'Taguig City': 8, 'Bras_lia': 9, 'Rio de Janeiro': 10, 'So Paulo': 11, 'Albany': 12, 'Armidale': 13, 'Athens': 14, 'Augusta': 15, 'Balingup': 16, 'Beechworth': 17, 'Boise': 18, 'Cedar Rapids/Iowa City': 19, 'Chatham-Kent': 20, 'Clatskanie': 21, 'Cochrane': 22, 'Columbus': 23, 'Consort': 24, 'Dalton': 25, 'Davenport': 26, 'Des Moines': 27, 'Dicky Beach': 28, 'Dubuque': 29, 'East Ballina': 30, 'Fernley': 31, 'Flaxton': 32, 'Forrest': 33, 'Gainesville': 34, 'Hepburn Springs': 35, 'Huskisson': 36, 'Inverloch': 37, 'Lakes Entrance': 38, 'Lakeview': 39, 'Lincoln': 40, 'Lorn': 41, 'Macedon': 42, 'Macon': 43, 'Mayfield': 44, 'Mc Millan': 45, 'Middleton Beach': 46, 'Miller': 47, 'Monroe': 48, 'Montville': 49, 'Ojo Caliente': 50,

In [116]:
cuisine_encoded.head() # merged at last

Afghani  African  American  Andhra  Arabian  Argentine  Armenian  Asian  \
0        0        0         0       0        0          0         0      0   
1        0        0         0       0        0          0         0      0   
2        0        0         0       0        0          0         0      1   
3        0        0         0       0        0          0         0      0   
4        0        0         0       0        0          0         0      0   

   Asian Fusion  Assamese  ...  Teriyaki  Tex-Mex  Thai  Tibetan  Turkish  \
0             0         0  ...         0        0     0        0        0   
1             0         0  ...         0        0     0        0        0   
2             0         0  ...         0        0     0        0        0   
3             0         0  ...         0        0     0        0        0   
4             0         0  ...         0        0     0        0        0   

   Turkish Pizza  Vegetarian  Vietnamese  Western  World Cuisine  
0              0           0           0        0              0  
1              0           0           0        0              0  
2              0           0           0        0              0  
3              0           0           0        0              0  
4              0           0           0        0              0  

[5 rows x 145 columns]

### Normalizing Numerical columns

In [117]:
for i in num:
    mean=round(data[i].mean(),2)
    sd=round(data[i].std(),2)
    new=[]
    for j in data[i]:
        z=((j-mean)/sd)
        new.append(z)
    data[i]=new
data.head()

Country Code  City  Average Cost for two  Currency  Has Table booking  \
0             0     0             -0.006154         0                  1   
1             0     0              0.000049         0                  1   
2             0     1              0.173734         0                  1   
3             0     1              0.018658         0                  0   
4             0     1              0.018658         0                  1   

   Has Online delivery  Is delivering now Price range  Aggregate rating  \
0                    0                  0           3          1.401316   
1                    0                  0           3          1.203947   
2                    0                  0           4          1.138158   
3                    0                  0           4          1.467105   
4                    0                  0           4          1.401316   

   Rating color  ...  Teriyaki  Tex-Mex  Thai  Tibetan  Turkish  \
0             0  ...         0        0     0        0        0   
1             0  ...         0        0     0        0        0   
2             1  ...         0        0     0        0        0   
3             0  ...         0        0     0        0        0   
4             0  ...         0        0     0        0        0   

   Turkish Pizza  Vegetarian  Vietnamese  Western  World Cuisine  
0              0           0           0        0              0  
1              0           0           0        0              0  
2              0           0           0        0              0  
3              0           0           0        0              0  
4              0           0           0        0              0  

[5 rows x 157 columns]

### Splitting the data into input and output 

In [118]:
X=data.drop('Aggregate rating',axis=1)
y=data['Aggregate rating']
X.head() ## Input columns

Country Code  City  Average Cost for two  Currency  Has Table booking  \
0             0     0             -0.006154         0                  1   
1             0     0              0.000049         0                  1   
2             0     1              0.173734         0                  1   
3             0     1              0.018658         0                  0   
4             0     1              0.018658         0                  1   

   Has Online delivery  Is delivering now Price range  Rating color  \
0                    0                  0           3             0   
1                    0                  0           3             0   
2                    0                  0           4             1   
3                    0                  0           4             0   
4                    0                  0           4             0   

   Rating text  ...  Teriyaki  Tex-Mex  Thai  Tibetan  Turkish  Turkish Pizza  \
0            0  ...         0        0     0        0        0              0   
1            0  ...         0        0     0        0        0              0   
2            1  ...         0        0     0        0        0              0   
3            0  ...         0        0     0        0        0              0   
4            0  ...         0        0     0        0        0              0   

   Vegetarian  Vietnamese  Western  World Cuisine  
0           0           0        0              0  
1           0           0        0              0  
2           0           0        0              0  
3           0           0        0              0  
4           0           0        0              0  

[5 rows x 156 columns]

In [119]:
y.head(10)#out put columns

0    1.401316
1    1.203947
2    1.138158
3    1.467105
4    1.401316
5    1.138158
6    0.875000
7    1.006579
8    1.467105
9    1.401316
Name: Aggregate rating, dtype: float64

### Linear Regression

In [120]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2, random_state=42)
X_train.shape,X_test.shape,y_train.shape,y_test.shape

((7640, 156), (1911, 156), (7640,), (1911,))

In [121]:
from sklearn.linear_model import LinearRegression
lr_model=LinearRegression()
lr_model

LinearRegression()

In [122]:
lr_model.fit(X_train,y_train)

LinearRegression()

In [123]:
y_pred=lr_model.predict(X_test)
pd.DataFrame(zip(y_test,y_pred),columns=['Y_actual','Y_pred']).head(10)

Y_actual    Y_pred
0 -0.375000 -1.931617
1  0.940789  1.446621
2  0.348684 -0.237238
3  1.138158  1.536247
4  0.546053  0.476131
5 -1.756579 -1.046343
6  0.348684  0.058219
7 -1.756579 -1.110232
8  0.611842  0.843992
9  0.875000  1.374748

In [124]:
mse=round((np.sum(np.square(y_test-y_pred)))/len(y_pred),3)
mse

2.8316217211979496e+16

### **3.1.2 Split the dataset into training and testing sets and evaluate the model's performance using appropriate metrics.**

In [125]:
from sklearn.metrics import mean_squared_error, r2_score
mse = mean_squared_error(y_test, y_pred)
rmse=mse**0.5
r2 = r2_score(y_test, y_pred)
print(f'Mean Squared Error: {mse}')
print(f'Root Mean Squared Error: {rmse}')
print(f'R-squared Score: {r2}')


Mean Squared Error: 2.8316217211979496e+16
Root Mean Squared Error: 168274232.16874143
R-squared Score: -2.874281878639827e+16


- **MSE=2.83×10^16**(extremely large error) means model's predictions outputs are far from the actual output values
- **R²=-2.8x10^16** means the model is performing worse than a simple mean model
- we can conclude that **Linear Regression Model's performance is very poor**

### **3.1.3 Experiment with different algorithms (e.g.linear regression, decision trees, random forest) and compare their performance.**


### DecisionTree

In [126]:
from sklearn.tree import DecisionTreeRegressor
dt_model=DecisionTreeRegressor()
dt_model

DecisionTreeRegressor()

In [127]:
dt_model.fit(X_train,y_train)

DecisionTreeRegressor()

In [128]:
y_pred=dt_model.predict(X_test)
pd.DataFrame(zip(y_test,y_pred),columns=['Y_actual','Y_pred']).head(10)

Y_actual    Y_pred
0 -0.375000 -0.243421
1  0.940789  0.875000
2  0.348684  0.085526
3  1.138158  0.940789
4  0.546053  0.546053
5 -1.756579 -1.756579
6  0.348684  0.414474
7 -1.756579 -1.756579
8  0.611842  0.611842
9  0.875000  0.875000

In [129]:
mse=round(np.sum(np.square(y_test-y_pred))/len(y_pred),4)
mse

0.0245

In [130]:
mse =round(mean_squared_error(y_test, y_pred),4)
rmse=round(mse**0.5,4)
r2 = round(r2_score(y_test, y_pred),4)
print(f'Mean Squared Error: {mse}')
print(f'Root Mean Squared Error: {rmse}')
print(f'R-squared Score: {r2}')

Mean Squared Error: 0.0245
Root Mean Squared Error: 0.1565
R-squared Score: 0.9751


- **As MSE=0.0239**(very low error) means model's predictions outputs are closer to the actual output values
- **R²=0.975** means the model explains 97.5% of the testing data
- So,**Decision Tree Regression model performance is Excellent**

### Random Forest

In [131]:
from sklearn.ensemble import RandomForestRegressor
rf_model=RandomForestRegressor()
rf_model

RandomForestRegressor()

In [132]:
rf_model.fit(X_train,y_train)

RandomForestRegressor()

In [133]:
y_pred=rf_model.predict(X_test)
pd.DataFrame(zip(y_test,y_pred),columns=['Y_actual','Y_pred']).head(10)

Y_actual    Y_pred
0 -0.375000 -0.219079
1  0.940789  0.978289
2  0.348684  0.176441
3  1.138158  0.938816
4  0.546053  0.555263
5 -1.756579 -1.756579
6  0.348684  0.221053
7 -1.756579 -1.756579
8  0.611842  0.619737
9  0.875000  0.928289

In [134]:
mse=round(np.sum(np.square(y_test-y_pred))/len(y_pred),4)
mse

0.0136

In [135]:
mse =round(mean_squared_error(y_test, y_pred),4)
rmse=round(mse**0.5,4)
r2 = round(r2_score(y_test, y_pred),4)
print(f'Mean Squared Error: {mse}')
print(f'Root Mean Squared Error: {rmse}')
print(f'R-squared Score: {r2}')

Mean Squared Error: 0.0136
Root Mean Squared Error: 0.1166
R-squared Score: 0.9862


- **As Mean Squared Error=0.0135**(very low error) means model's predictions outputs are closer to the actual output values
- **R-squared Score=0.9863** means the model explains 98.63% of the testing data
- So,**Random Forest` Regression model performance is Excellent**
- #### **Comparing to Linear Regression,Decision Tree&Random Forest we can say that Random forest model for the above dataset**